moe/ucw/doc/config.txt
Configuration files
===================

This document describes run-time configuration of libucw-based
programs using config files. For compile-time configuration,
see <<configure:>>.

[[terminology]]
Terminology
-----------

Configuration items of all modules are organized into sections.
The sections form a tree structure with top-level sections corresponding
to program modules.

Each configuration item belongs to one of the following classes:

  1. single value or a fixed-length array of values
  2. variable-length array of values
  3. subsection with several nested attributes
  4. list of nodes, each being an instance of a subsection
  5. bitmap of small integers (0..31) or fixed list of strings
  6. exceptions (items with irregular syntax; however, they always
     appear as a sequence of strings, only the semantics differ)

Both fixed- and variable-length arrays consist of items of the same
type.  The basic types supported by the configuration mechanism are:

  1. 32-bit integer
  2. 64-bit integer
  3. floating point number
  4. IP address
  5. string
  6. choice (one of a fixed list of strings)

Program modules can define their own special types (such as network
masks or attribute names) and decide how are they parsed.

[[format]]
Format of configuration files
-----------------------------

Configuration files are text files that usually set one attribute per
line, though it is possible to split one assignment into multiple lines
and/or assign several attributes in one line.  The basic format of an
assignment command is

  name value1 value2 ... valueN

or

  name=value1 value2 ... valueN

The end of line means also end of a command unless it is preceded by a
backslash.  On the other hand, a semicolon terminates the command and
another command can start after the semicolon.  A hash starts a comment
that lasts until the end of the line.  A value can be enclosed in
apostrophes or quotation marks and then it can contain spaces and/or
control characters, otherwise the first space or control character
denotes the end of the value.  Values enclosed in quotation marks are
interpreted as C-strings.  For example, the following are valid
assignment commands:

  Database "main db\x2b"; Directory='index/'; Weights 100 20 30 \
    40 50 80				# a comment that is ignored

Numerical values can be succeeded by a unit.  The following units are
supported:

[[units]]

	d=86400		k=1000		K=1024
	h=3600		m=1000000	M=1048576
	%=0.01		g=1000000000	G=1073741824

Attributes of a section or a list node can be set in two ways.  First,
you can write the name of the section or list, open a bracket, and then
set the attributes inside the section.  For example,

  Section1 {
    Attr1	value1
    Attr2	value2
    ListNode {		#creates a list and adds its first node
      Attr3	value3
      Attr4	value4
    }
    ListNode { Attr3=value5; Attr4=value6 }
			#appends a new node; this is still the same syntax
  }

The second possibility is using a shorter syntax when all attributes of a
section are set on one line in a fixed order.  The above example could
be as well written as

  Section1 {
    Attr1	value1
    Attr2	value2
    ListNode	value3 value4
    ListNode	value5 value6
  }

Of course, you cannot use the latter syntax when the attributes allow
variable numbers of parameters.  The parser of the configuration files
checks this possibility.

If you want to set a single attribute in some section, you can also
refer to the attribute as Section.Attribute.

Lists support several operations besides adding a new node.  You just
have to write a colon immediately after the attribute name, followed by
the name of the operation.  The following operations are supported:

[[operations]]

  List:clear				# removes all nodes
  List:append { attr1=value1; ... }	# adds a new node at the end
  List:prepend { attr1=value1; ... }	# adds a new node at the beginning
  List:remove { attr1=search1 }		# find a node and delete it
  List:edit { attr1=search1 } { attr1=value1; ... }
					# find a node and edit it
  List:after { attr1=search1 } { ... }	# insert a node after a found node
  List:before { attr1=search1 } { ... }	# insert a node before a found node
  List:copy { attr1=search1 } { ... }	# duplicate a node and edit the copy
  List:reset { attr=value1; ... }	# equivalent to :clear and :append

You can specify several attributes in the search condition and the nodes
are tested for equality in all these attributes.  In the editing
commands, you can either open a second block with overridden attributes,
or specify the new values using the shorter one-line syntax.

The commands :clear, :append, and :prepend are also supported by var-length
arrays.  The command :clear can also be used on string values.  The following
operations can be used on bitmaps: :set (which is equal to :append and :prepend),
:remove, :clear, and :all (set all bits).

[[include]]
Including other files
---------------------

To include another file, use the command

  include another/file

(Beware that this command has to be the last one on the line.)

[[command_line]]
Command-line parameters
-----------------------

The default configuration file (cf_def_file possibly overriden
by environment variable cf_env_file) is read before the program is started.
You can use a -C option to override the name of the configuration file.
If you use this parameter several times, then all those files are loaded
consecutively. A parameter -S can be used to execute a configuration
command directly (after loading the default or specified configuration
file).  Example:

  bin/program -Ccf/my-config -S'module.trace=2;module.logfile:clear' ...

If the program is compiled with debugging information, then one more
parameter `--dumpconfig` is supported.  It prints all parsed configuration
items and exits.

All these switches must be used before any other parameters of the
program.

[[preprocess]]
Preprocessing
-------------

During compilation, all configuration files are pre-processed by a simple
C-like preprocessor, which supports `#ifdef`, `#ifndef`, `#if`,
`#elsif`, `#else` and `#endif` directives referring to compile-time
configuration variables (the ones detected by `configure` script, you
can see list of them in `obj/autoconf.h`). `#if` and `#elsif` can contain
any Perl expression where each `CONFIG_xyz` configuration variable is
substituted to 0 or 1 depending on its value.

The preprocessor also substitutes `@VARIABLE@` by the value of the variable,
which must be defined.

[[caveats]]
Caveats
-------

Trying to access an unknown attribute causes an error, but unrecognized
top-level sections are ignored.  The reason is that a common config file
is used for a lot of programs which recognize only their own sections.

Names of sections, attributes and choices are case-insensitive. Units are
case-sensitive.


moe/ucw/doc/configure.txt
How to Configure Sherlock libraries
===================================

What can be configured
----------------------
There are two different levels of configuring/customizing program
based on sherlock libraries:

  - runtime configuration in configuration files (see <<config:>>)

  - compile-time configuration of the libraries: config switches set
    before compiling, selecting optional features.

Where to build
--------------
If you run configure in the source directory, it prepares for compilation inside
the source tree. In this case, an `obj` subdirectory is created to hold all generated
files (object files, binaries, generated source files etc.) and all final files
are linked to the `run` subdirectory. No other parts of the source tree are written into.

Alternatively, you can compile in a separate object tree (which is useful when you
want to build several different configurations from a single source tree). In order
to do that, switch to the destination directory and issue `<source-dir>/configure ...`.
This way, configure will create the `obj` and `run` directories locally and set up
a Makefile which refers to the original source tree.

How to configure
----------------
To set up compilation, possibly overriding default compile-time
options, just run:

	./configure [<option> | -<option> | <option>=<value> ...]

The default values of feature options are taken from `default.cfg`.
Compiler flags and options dependent on compiler, OS and CPU type
are set in `ucw/perl/UCW/Configure/C.pm`. Everything can be overriden by
options specified on the configure's command line, which have the highest
priority.

If you want to see the resulting set of options, take a look at
`obj/config.mk`.

Options specifying compiler/linker/debugger options can be also overriden
during compilation by `make <option>=<value>`. While it's also possible
to specify the other options in this way, it probably won't have the desired
effect, because configure also generates C include files containing the
options.

Installation options
--------------------
By default, the package is compiled to be installed and it chooses a
place where it will install. If you want to place it somewhere else,
set the `PREFIX` option:

	./configure PREFIX=/where/to/install

[[local_build]]
You can create a local compilation, which does not need to be
installed. All paths are set relative to a `run` directory and
programs expect to have current working directory set to it. However,
you can move the `run` directory around your filesystem as you like.
To compile it that way, turn on the `CONFIG_LOCAL` option (and do not
set `PREFIX`):

	./configure CONFIG_LOCAL

Examples
--------
  - `./configure -CONFIG_SHARED` will build the libraries statically.
  - `./configure -CONFIG_IMAGES` avoids building the image library.
  - `./configure CONFIG_UCW_ONLY -CONFIG_UCW_UTILS` compiles only the
    libucw library without its util programs.
  - `./configure CONFIG_XML PREFIX=$HOME/sherlock` includes a xml
    library and installs into your home directory.


moe/ucw/conf.h
/*
 *	UCW Library -- Configuration files
 *
 *	(c) 2001--2006 Robert Spalek <robert@ucw.cz>
 *	(c) 2003--2006 Martin Mares <mj@ucw.cz>
 *
 *	This software may be freely distributed and used according to the terms
 *	of the GNU Lesser General Public License.
 */

#ifndef	_UCW_CONF_H
#define	_UCW_CONF_H

/*** === Data types [[conf_types]] ***/

enum cf_class {				/** Class of the configuration item. **/
  CC_END,				// end of list
  CC_STATIC,				// single variable or static array
  CC_DYNAMIC,				// dynamically allocated array
  CC_PARSER,				// arbitrary parser function
  CC_SECTION,				// section appears exactly once
  CC_LIST,				// list with 0..many nodes
  CC_BITMAP				// of up to 32 items
};

enum cf_type {				/** Type of a single value. **/
  CT_INT, CT_U64, CT_DOUBLE,		// number types
  CT_IP,				// IP address
  CT_STRING,				// string type
  CT_LOOKUP,				// in a string table
  CT_USER				// user-defined type
};

struct fastbuf;

/**
 * A parser function gets an array of (strdup'ed) strings and a pointer with
 * the customized information (most likely the target address).  It can store
 * the parsed value anywhere in any way it likes, however it must first call
 * @cf_journal_block() on the overwritten memory block.  It returns an error
 * message or NULL if everything is all right.
 **/
typedef char *cf_parser(uns number, char **pars, void *ptr);
/**
 * A parser function for user-defined types gets a string and a pointer to
 * the destination variable.  It must store the value within [ptr,ptr+size),
 * where size is fixed for each type.  It should not call @cf_journal_block().
 **/
typedef char *cf_parser1(char *string, void *ptr);
/**
 * An init- or commit-hook gets a pointer to the section or NULL if this
 * is the global section.  It returns an error message or NULL if everything
 * is all right.  The init-hook should fill in default values (needed for
 * dynamically allocated nodes of link lists or for filling global variables
 * that are run-time dependent).  The commit-hook should perform sanity
 * checks and postprocess the parsed values.  Commit-hooks must call
 * @cf_journal_block() too.  Caveat! init-hooks for static sections must not
 * use @cf_malloc() but normal <<memory:xmalloc()>>.
 **/
typedef char *cf_hook(void *ptr);
/**
 * Dumps the contents of a variable of a user-defined type.
 **/
typedef void cf_dumper1(struct fastbuf *fb, void *ptr);
/**
 * Similar to init-hook, but it copies attributes from another list node
 * instead of setting the attributes to default values.  You have to provide
 * it if your node contains parsed values and/or sub-lists.
 **/
typedef char *cf_copier(void *dest, void *src);

struct cf_user_type {			/** Structure to store information about user-defined variable type. **/
  uns size;				// of the parsed attribute
  char *name;				// name of the type (for dumping)
  cf_parser1 *parser;			// how to parse it
  cf_dumper1 *dumper;			// how to dump the type
};

struct cf_section;
struct cf_item {			/** Single configuration item. **/
  const char *name;			// case insensitive
  int number;				// length of an array or #parameters of a parser (negative means at most)
  void *ptr;				// pointer to a global variable or an offset in a section
  union cf_union {
    struct cf_section *sec;		// declaration of a section or a list
    cf_parser *par;			// parser function
    const char * const *lookup;		// NULL-terminated sequence of allowed strings for lookups
    struct cf_user_type *utype;		// specification of the user-defined type
  } u;
  enum cf_class cls:16;			// attribute class
  enum cf_type type:16;			// type of a static or dynamic attribute
};

struct cf_section {			/** A section. **/
  uns size;				// 0 for a global block, sizeof(struct) for a section
  cf_hook *init;			// fills in default values (no need to bzero)
  cf_hook *commit;			// verifies parsed data (optional)
  cf_copier *copy;			// copies values from another instance (optional, no need to copy basic attributes)
  struct cf_item *cfg;			// CC_END-terminated array of items
  uns flags;				// for internal use only
};

/***
 * [[conf_macros]]
 * Convenience macros
 * ~~~~~~~~~~~~~~~~~~
 *
 * You could create the structures manually, but you can use these macros to
 * save some typing.
 */

/***
 * Declaration of <<struct_cf_section,`cf_section`>>
 * ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 *
 * These macros can be used to configure the <<struct_cf_section,`cf_section`>>
 * structure.
 ***/

/**
 * Data type of a section.
 * If you store the section into a structure, use this macro.
 *
 * Storing a section into a structure is useful mostly when you may have multiple instances of the
 * section (eg. <<conf_multi,array or list>>).
 *
 * Example:
 *
 *   struct list_node {
 *     cnode n;		// This one is for the list itself
 *     char *name;
 *     uns value;
 *   };
 *
 *   static struct clist nodes;
 *
 *   static struct cf_section node = {
 *     CF_TYPE(struct list_node),
 *     CF_ITEMS {
 *       CF_STRING("name", PTR_TO(struct list_node, name)),
 *       CF_UNS("value", PTR_TO(struct list_node, value)),
 *       CF_END
 *     }
 *   };
 *
 *   static struct cf_section section = {
 *     CF_LIST("node", &nodes, &node),
 *     CF_END
 *   };
 *
 * You could use <<def_CF_STATIC,`CF_STATIC`>> or <<def_CF_DYNAMIC,`CF_DYNAMIC`>>
 * macros to create arrays.
 */
#define CF_TYPE(s)	.size = sizeof(s)
/**
 * An init <<hooks,hook>>.
 * You can use this to initialize dynamically allocated items (for a dynamic array or list).
 * The hook returns an error message or NULL if everything was OK.
 */
#define CF_INIT(f)	.init = (cf_hook*) f
/**
 * A commit <<hooks,hook>>.
 * You can use this one to check sanity of loaded data and postprocess them.
 * You must call @cf_journal_block() if you change anything.
 *
 * Return error message or NULL if everything went OK.
 **/
#define CF_COMMIT(f)	.commit = (cf_hook*) f
/**
 * A <<hooks,copy function>>.
 * You need to provide one for too complicated sections where a memcpy is not
 * enough to copy it properly. It happens, for example, when you have a dynamically
 * allocated section containing a list of other sections.
 *
 * You return an error message or NULL if you succeed.
 **/
#define CF_COPY(f)	.copy = (cf_copier*) f		/**  **/
#define CF_ITEMS	.cfg = ( struct cf_item[] )	/** List of sub-items. **/
#define CF_END		{ .cls = CC_END }		/** End of the structure. **/
/***
 * Declaration of a configuration item
 * ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 *
 * Each of these describe single <<struct_cf_item,configuration item>>. They are mostly
 * for internal use, do not use them directly unless you really know what you are doing.
 ***/

/**
 * Static array of items.
 * Expects you to allocate the memory and provide pointer to it.
 **/
#define CF_STATIC(n,p,T,t,c)	{ .cls = CC_STATIC, .type = CT_##T, .name = n, .number = c, .ptr = CHECK_PTR_TYPE(p,t*) }
/**
 * Dynamic array of items.
 * Expects you to provide pointer to your pointer to data and it will allocate new memory for it
 * and set your pointer to it.
 **/
#define CF_DYNAMIC(n,p,T,t,c)	{ .cls = CC_DYNAMIC, .type = CT_##T, .name = n, .number = c, .ptr = CHECK_PTR_TYPE(p,t**) }
#define CF_PARSER(n,p,f,c)	{ .cls = CC_PARSER, .name = n, .number = c, .ptr = p, .u.par = (cf_parser*) f }					/** A low-level parser. **/
#define CF_SECTION(n,p,s)	{ .cls = CC_SECTION, .name = n, .number = 1, .ptr = p, .u.sec = s }						/** A sub-section. **/
#define CF_LIST(n,p,s)		{ .cls = CC_LIST, .name = n, .number = 1, .ptr = CHECK_PTR_TYPE(p,clist*), .u.sec = s }				/** A list with sub-items. **/
#define CF_BITMAP_INT(n,p)	{ .cls = CC_BITMAP, .type = CT_INT, .name = n, .number = 1, .ptr = CHECK_PTR_TYPE(p,u32*) }			/** A bitmap. **/
#define CF_BITMAP_LOOKUP(n,p,t)	{ .cls = CC_BITMAP, .type = CT_LOOKUP, .name = n, .number = 1, .ptr = CHECK_PTR_TYPE(p,u32*), .u.lookup = t }	/** A bitmap with named bits. **/
/***
 * Basic configuration items
 * ^^^^^^^^^^^^^^^^^^^^^^^^^
 *
 * They describe basic data types used in the configuration. This should be enough for
 * most real-life purposes.
 *
 * The parameters are as follows:
 *
 * * @n -- name of the item.
 * * @p -- pointer to the variable where it shall be stored.
 * * @c -- count.
 **/
#define CF_INT(n,p)		CF_STATIC(n,p,INT,int,1)		/** Single `int` value. **/
#define CF_INT_ARY(n,p,c)	CF_STATIC(n,p,INT,int,c)		/** Static array of integers. **/
#define CF_INT_DYN(n,p,c)	CF_DYNAMIC(n,p,INT,int,c)		/** Dynamic array of integers. **/
#define CF_UNS(n,p)		CF_STATIC(n,p,INT,uns,1)		/** Single `uns` (`unsigned`) value. **/
#define CF_UNS_ARY(n,p,c)	CF_STATIC(n,p,INT,uns,c)		/** Static array of unsigned integers. **/
#define CF_UNS_DYN(n,p,c)	CF_DYNAMIC(n,p,INT,uns,c)		/** Dynamic array of unsigned integers. **/
#define CF_U64(n,p)		CF_STATIC(n,p,U64,u64,1)		/** Single unsigned 64bit integer (`u64`). **/
#define CF_U64_ARY(n,p,c)	CF_STATIC(n,p,U64,u64,c)		/** Static array of u64s. **/
#define CF_U64_DYN(n,p,c)	CF_DYNAMIC(n,p,U64,u64,c)		/** Dynamic array of u64s. **/
#define CF_DOUBLE(n,p)		CF_STATIC(n,p,DOUBLE,double,1)		/** Single instance of `double`. **/
#define CF_DOUBLE_ARY(n,p,c)	CF_STATIC(n,p,DOUBLE,double,c)		/** Static array of doubles. **/
#define CF_DOUBLE_DYN(n,p,c)	CF_DYNAMIC(n,p,DOUBLE,double,c)		/** Dynamic array of doubles. **/
#define CF_IP(n,p)		CF_STATIC(n,p,IP,u32,1)			/** Single IPv4 address. **/
#define CF_IP_ARY(n,p,c)	CF_STATIC(n,p,IP,u32,c)			/** Static array of IP addresses. **/.
#define CF_IP_DYN(n,p,c)	CF_DYNAMIC(n,p,IP,u32,c)		/** Dynamic array of IP addresses. **/
/**
 * A string.
 * You provide a pointer to a `char *` variable and it will fill it with
 * dynamically allocated string. For example:
 *
 *   static char *string = "Default string";
 *
 *   static struct cf_section section = {
 *     CF_ITEMS {
 *       CF_STRING("string", &string),
 *       CF_END
 *     }
 *   };
 **/
#define CF_STRING(n,p)		CF_STATIC(n,p,STRING,char*,1)
#define CF_STRING_ARY(n,p,c)	CF_STATIC(n,p,STRING,char*,c)		/** Static array of strings. **/
#define CF_STRING_DYN(n,p,c)	CF_DYNAMIC(n,p,STRING,char*,c)		/** Dynamic array of strings. **/
/**
 * One string out of a predefined set.
 * You provide the set as an array of strings terminated by NULL (similar to @argv argument
 * of main()) as the @t parameter.
 *
 * The configured variable (pointer to `int`) is set to index of the string.
 * So, it works this way:
 *
 *   static *strings[] = { "First", "Second", "Third", NULL };
 *
 *   static int variable;
 *
 *   static struct cf_section section = {
 *     CF_ITEMS {
 *       CF_LOOKUP("choice", &variable, strings),
 *       CF_END
 *     }
 *   };
 *
 * Now, if the configuration contains `choice "Second"`, `variable` will be set to 1.
 **/
#define CF_LOOKUP(n,p,t)	{ .cls = CC_STATIC, .type = CT_LOOKUP, .name = n, .number = 1, .ptr = CHECK_PTR_TYPE(p,int*), .u.lookup = t }
/**
 * Static array of strings out of predefined set.
 **/
#define CF_LOOKUP_ARY(n,p,t,c)	{ .cls = CC_STATIC, .type = CT_LOOKUP, .name = n, .number = c, .ptr = CHECK_PTR_TYPE(p,int*), .u.lookup = t }
/**
 * Dynamic array of strings out of predefined set.
 **/
#define CF_LOOKUP_DYN(n,p,t,c)	{ .cls = CC_DYNAMIC, .type = CT_LOOKUP, .name = n, .number = c, .ptr = CHECK_PTR_TYPE(p,int**), .u.lookup = t }
/**
 * A user-defined type.
 * See <<custom_parser,creating custom parsers>> section if you want to know more.
 **/
#define CF_USER(n,p,t)		{ .cls = CC_STATIC, .type = CT_USER, .name = n, .number = 1, .ptr = p, .u.utype = t }
/**
 * Static array of user-defined types (all of the same type).
 * See <<custom_parser,creating custom parsers>> section.
 **/
#define CF_USER_ARY(n,p,t,c)	{ .cls = CC_STATIC, .type = CT_USER, .name = n, .number = c, .ptr = p, .u.utype = t }
/**
 * Dynamic array of user-defined types.
 * See <<custom_parser,creating custom parsers>> section.
 **/
#define CF_USER_DYN(n,p,t,c)	{ .cls = CC_DYNAMIC, .type = CT_USER, .name = n, .number = c, .ptr = p, .u.utype = t }

/**
 * Any number of dynamic array elements
 **/
#define CF_ANY_NUM		-0x7fffffff

#define DARY_LEN(a) ((uns*)a)[-1]	/** Length of an dynamic array. **/
#define DARY_ALLOC(type,len,val...) ((struct { uns l; type a[len]; }) { .l = len, .a = { val } }).a
  // creates a static instance of a dynamic array

/***
 * [[alloc]]
 * Memory allocation
 * ~~~~~~~~~~~~~~~~~
 *
 * Uses <<mempool:,memory pools>> for efficiency and journal recovery.
 * You should use these routines when implementing custom parsers.
 ***/
struct mempool;
/**
 * A <<mempool:type_mempool,memory pool>> for configuration parser needs.
 * Memory allocated from here is valid as long as the current config is loaded
 * (if you allocate some memory and rollback the transaction or you load some
 * other configuration, it gets lost).
 **/
extern struct mempool *cf_pool;
void *cf_malloc(uns size);	/** Returns @size bytes of memory. Allocates from <<var_cf_pool,`cf_pool`>>. **/
void *cf_malloc_zero(uns size);	/** Like @cf_malloc(), but zeroes the memory. **/
char *cf_strdup(const char *s);	/** Copy a string into @cf_malloc()ed memory. **/
char *cf_printf(const char *fmt, ...) FORMAT_CHECK(printf,1,2); /** printf() into @cf_malloc()ed memory. **/

/***
 * [[journal]]
 * Undo journal
 * ~~~~~~~~~~~~
 *
 * For error recovery when <<reload,reloading configuration>>.
 ***/
extern uns cf_need_journal;	/** Is the journal needed? If you do not reload configuration, you set this to 0 and gain a little more performance and free memory. **/
/**
 * When a block of memory is about to be changed, put the old value
 * into journal with this function. You need to call it from a <<hooks,commit hook>>
 * if you change anything. It is used internally by low-level parsers.
 * <<custom_parser,Custom parsers>> do not need to call it, it is called
 * before them.
 **/
void cf_journal_block(void *ptr, uns len);
#define CF_JOURNAL_VAR(var) cf_journal_block(&(var), sizeof(var))	// Store single value into journal.

/***
 * [[declare]]
 * Section declaration
 * ~~~~~~~~~~~~~~~~~~~
 **/

/**
 * Plug another top-level section into the configuration system.
 * @name is the name in the configuration file,
 * @sec is pointer to the section description.
 * If @allow_unknown is set to 0 and a variable not described in @sec
 * is found in the configuration file, it produces an error.
 * If you set it to 1, all such variables are ignored.
 **/
void cf_declare_section(const char *name, struct cf_section *sec, uns allow_unknown);
/**
 * If you have a section in a structure and you want to initialize it
 * (eg. if you want a copy of default values outside the configuration),
 * you can use this. It initializes it recursively.
 *
 * This is used mostly internally. You probably do not need it.
 **/
void cf_init_section(const char *name, struct cf_section *sec, void *ptr, uns do_bzero);

/***
 * [[bparser]]
 * Parsers for basic types
 * ~~~~~~~~~~~~~~~~~~~~~~~
 *
 * Each of them gets a string to parse and pointer to store the value.
 * It returns either NULL or error message.
 *
 * The parsers support units. See <<config:units,their list>>.
 ***/
char *cf_parse_int(const char *str, int *ptr);		/** Parser for integers. **/
char *cf_parse_u64(const char *str, u64 *ptr);		/** Parser for 64 unsigned integers. **/
char *cf_parse_double(const char *str, double *ptr);	/** Parser for doubles. **/
char *cf_parse_ip(const char *p, u32 *varp);		/** Parser for IP addresses. **/

#endif

